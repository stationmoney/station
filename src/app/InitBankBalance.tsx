import { PropsWithChildren } from "react"
import {
  useInitialBankBalance,
  useInitialTokenBalance,
} from "data/queries/bank"
import { BankBalanceProvider } from "data/queries/bank"
import NetworkLoading from "./NetworkLoading"

const InitBankBalance = ({ children }: PropsWithChildren<{}>) => {
  const { data: bankBalance } = useInitialBankBalance()
  const { data: tokenBalance } = useInitialTokenBalance()
  // If the balance doesn't exist, nothing is worth rendering.
  if (!bankBalance) return <NetworkLoading title="Fetching balances..." />
  return (
    <BankBalanceProvider value={[...bankBalance, ...(tokenBalance ?? [])]}>
      {children}
    </BankBalanceProvider>
  )
}

export default InitBankBalance
