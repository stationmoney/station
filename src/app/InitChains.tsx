import axios from "axios"
import { ASSETS } from "config/constants"
import { WhitelistProvider } from "data/queries/chains"
import { PropsWithChildren, useEffect, useState } from "react"

type Whitelist = Record<
  string,
  {
    token: string
    symbol: string
    name: string
    icon: string
    chains: string[]
    decimals: number
  }
>

const InitChains = ({ children }: PropsWithChildren<{}>) => {
  const [data, setData] = useState<Whitelist>()
  useEffect(() => {
    ;(async () => {
      const { data: whitelist } = await axios.get("/station/coins.json", {
        baseURL: ASSETS,
      })
      setData(whitelist)
    })()
  }, [])

  if (!data) return null

  return <WhitelistProvider value={data}>{children}</WhitelistProvider>
}

export default InitChains
