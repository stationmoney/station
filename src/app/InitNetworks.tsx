import { PropsWithChildren, useEffect, useState } from "react"
import axios from "axios"
import { ASSETS } from "config/constants"
import createContext from "utils/createContext"
import NetworkLoading from "./NetworkLoading"

export const [useNetworks, NetworksProvider] = createContext<{
  networks: InterchainNetworks
  filterEnabledNetworks: <T>(network: Record<string, T>) => Record<string, T>
}>("useNetworks")

const InitNetworks = ({ children }: PropsWithChildren<{}>) => {
  const [networks, setNetworks] = useState<InterchainNetworks>()
  const [enabledNetworks, setEnabledNetworks] = useState<string[]>([])

  useEffect(() => {
    const fetchChains = async () => {
      const { data: chains } = await axios.get<InterchainNetworks>(
        "/station/chains.json",
        {
          baseURL: ASSETS,
        }
      )
      setNetworks(chains)
    }

    fetchChains()
  }, [])

  useEffect(() => {
    const testChains = async () => {
      if (!networks) return
      const testBase = {
        ...networks.mainnet,
        ...networks.testnet,
        ...networks.classic,
      }

      const result = await Promise.all(
        Object.values(testBase).map(async (network) => {
          try {
            const { data } = await axios.get(
              "cosmos/base/tendermint/v1beta1/node_info",
              {
                baseURL: network.lcd,
                timeout: 5_000,
              }
            )
            return data.default_node_info.network as string
          } catch (e) {
            return null
          }
        })
      )

      setEnabledNetworks(
        result.filter((r) => typeof r === "string") as string[]
      )
    }

    testChains()
  }, [networks])

  if (!networks || !enabledNetworks.length)
    return <NetworkLoading title="Connecting to available networks..." />

  return (
    <NetworksProvider
      value={{
        networks,
        filterEnabledNetworks: (networks) =>
          Object.fromEntries(
            Object.entries(networks).filter(([chainID]) =>
              enabledNetworks.includes(chainID)
            )
          ),
      }}
    >
      {children}
    </NetworksProvider>
  )
}

export default InitNetworks
