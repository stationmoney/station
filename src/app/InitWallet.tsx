import { PropsWithChildren, useEffect, useMemo } from "react"
import { QueryClient, QueryClientProvider } from "react-query"
import { WalletStatus } from "@terra-money/wallet-types"
import { useWallet } from "@terra-money/use-wallet"
import { useNetworkName } from "data/wallet"
import { isWallet, useAuth } from "auth"
import Online from "./containers/Online"
import NetworkLoading from "./NetworkLoading"

const InitWallet = ({ children }: PropsWithChildren<{}>) => {
  useOnNetworkChange()
  const { status } = useWallet()
  const queryClient = useQueryClient()
  const networkName = useNetworkName()

  return status === WalletStatus.INITIALIZING ? (
    <NetworkLoading title="Initializing your wallet..." />
  ) : (
    <QueryClientProvider client={queryClient} key={networkName}>
      {children}
      <Online />
    </QueryClientProvider>
  )
}

export default InitWallet

/* hooks */
const useOnNetworkChange = () => {
  const { wallet, disconnect } = useAuth()
  const isPreconfiguredWallet = isWallet.preconfigured(wallet)
  const shouldDisconnect = isPreconfiguredWallet

  useEffect(() => {
    if (shouldDisconnect) disconnect()
  }, [disconnect, shouldDisconnect])
}

const useQueryClient = () => {
  const name = useNetworkName()

  return useMemo(() => {
    if (!name) throw new Error()
    return new QueryClient()
  }, [name])
}
