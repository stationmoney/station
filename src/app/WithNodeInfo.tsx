import { PropsWithChildren } from "react"
import { useNodeInfo } from "data/queries/tendermint"
import Overlay from "./components/Overlay"
import NetworkError from "./NetworkError"
import { useNetworkName } from "data/wallet"

const WithNodeInfo = ({ children }: PropsWithChildren<{}>) => {
  const network = useNetworkName()
  // FIXME: remove or make sure chainID is available
  const { isLoading, isError } = useNodeInfo(
    network === "mainnet" ? "phoenix-1" : "pisco-1"
  )

  if (isError) {
    return (
      <Overlay>
        <NetworkError />
      </Overlay>
    )
  }

  if (isLoading) return null
  return <>{children}</>
}

export default WithNodeInfo
