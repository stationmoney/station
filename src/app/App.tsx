import { getErrorMessage } from "utils/error"
import Layout, { MainContainer, Page } from "components/layout"
import { Banner, Content, Header, Actions, Sidebar } from "components/layout"
import { ErrorBoundary, Wrong } from "components/feedback"

/* routes */
import { useNav } from "./routes"

/* banner */
import UpdateExtension from "./sections/UpdateExtension"

/* sidebar */
import Nav from "./sections/Nav"
import Aside from "./sections/Aside"

/* header */
import NetworkHeader from "./sections/NetworkHeader"
import Refresh from "./sections/Refresh"
import Preferences from "./sections/Preferences"
import ConnectWallet from "./sections/ConnectWallet"

/* extra */
import LatestTx from "./sections/LatestTx"
import ValidatorButton from "./sections/ValidatorButton"
import DevTools from "./sections/DevTools"
import { useShowWelcomeModal } from "utils/localStorage"

/* init */
import InitBankBalance from "./InitBankBalance"
import Wallet from "pages/wallet/Wallet"
import WelcomeModal from "./sections/WelcomeModal"
import NavButton from "./sections/NavButton"
import NetworkStatus from "components/display/NetworkStatus"

const App = () => {
  const { element: routes } = useNav()
  const showModal = useShowWelcomeModal()

  return (
    <Layout>
      <Banner>
        <UpdateExtension />
      </Banner>

      <Sidebar>
        <Nav />
        <Aside />
      </Sidebar>

      <Header>
        <NetworkHeader />

        <Actions>
          <DevTools />
          <section>
            <Refresh />
            <Preferences />
            <NetworkStatus />
          </section>
          <ValidatorButton />
          <ConnectWallet />
          <NavButton />
        </Actions>
        <LatestTx />
      </Header>

      <Content>
        <ErrorBoundary fallback={fallback}>
          <InitBankBalance>
            <MainContainer>
              {routes}
              <Wallet />
              {showModal && <WelcomeModal />}
            </MainContainer>
          </InitBankBalance>
        </ErrorBoundary>
      </Content>
    </Layout>
  )
}

export default App

/* error */
export const fallback = (error: Error) => (
  <Page>
    <Wrong>{getErrorMessage(error)}</Wrong>
  </Page>
)
