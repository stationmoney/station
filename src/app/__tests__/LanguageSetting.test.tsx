import React from "react"
import { render, screen } from "@testing-library/react"
import LanguageSetting from "../sections/LanguageSetting"
import { NetworksProvider } from "app/InitNetworks"
import user from "@testing-library/user-event"

function renderComponent() {
  type TokenFilter = <T>(network: Record<string, T>) => Record<string, T>
  const networks = {} as jest.Mocked<InterchainNetworks>
  const mockedTokenFilter = {} as jest.Mocked<TokenFilter>

  return render(
    <NetworksProvider
      value={{
        networks: networks,
        filterEnabledNetworks: mockedTokenFilter,
        filterDisabledNetworks: mockedTokenFilter,
      }}
    >
      <LanguageSetting />
    </NetworksProvider>
  )
}

describe("LanguageSetting component matches snapshots", () => {
  it("matches original component", () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })

  it("matches user augmented component", async () => {
    const { asFragment } = renderComponent()

    // Change language to Chinese.
    const chineseLanguage = screen.getByRole("button", {
      name: "中文",
    })
    await user.click(chineseLanguage)

    // Ensure snapshot matches user augmented component.
    expect(asFragment()).toMatchSnapshot()
  })
})
