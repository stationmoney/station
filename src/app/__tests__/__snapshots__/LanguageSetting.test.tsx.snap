// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSetting component matches snapshots matches original component 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <section
      class="selector"
    >
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Deutsch
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              English
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Español
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Français
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Italiano
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Polski
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              中文
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`LanguageSetting component matches snapshots matches user augmented component 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <section
      class="selector"
    >
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Deutsch
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              English
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Español
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Français
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Italiano
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Polski
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              中文
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
    </section>
  </div>
</DocumentFragment>
`;
